### PR TITLE
Rework cdba to support controlling device via external program

### PR DESCRIPTION
**Pull request by lumag**
Here are the changes that I have here to allow controlling devices through the external script.

### CODE REVIEW

**Comment by lumag** [comment]
@andersson gracious ping

**Comment by lumag** [comment]
Rebased on top of current master branch.
@andersson any hope to hear response from you?

File: external.c [hunk 1]
```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2021, Linaro Ltd.
```
Comment (superna9999) [inline external.c:2]: Should be 2023

File: device.h [hunk 2]
```diff
@@ -60,6 +60,9 @@ struct device_user {
 	struct list_head node;
 };
 
+#define device_has_control(_dev, _op) \
```
Comment (superna9999) [inline device.h:63]: This check macro is nice but I think the corresponding getter macro would be nice to make code calling ops simpler

**Review by superna9999 — COMMENTED** [review]
The last patch fixes the external support or fixes in all cases ? If it fixes all cases it should got in a separate or right ?

File: external.c (hunk 1 above)
Comment (lumag) [reply external.c:2]: No. The patches were first published in 2021.

Comment (superna9999) [reply external.c:2]: Yeah but you reworked them today so at least add 2021-2023

Comment (lumag) [reply external.c:2]: sounds reasonable

**Comment by lumag** [comment]
> The last patch fixes the external support or fixes in all cases ? If it fixes all cases it should got in a separate or right ?

It is a generic fix, but I stumbled upon it only with the "external" driver, since other drivers do not output a lot.

File: device.h (hunk 2 above)
Comment (lumag) [reply device.h]: I have implemented a slightly different approach. Let me know what do you think about it.

**Comment by calebccff** [comment]
Is it possible to combine this with the qcomlt debug board? For the phones I have a Pi Pico exposing a compatible serial control interface, but for USB VBUS I need to use PPPS control...

**Comment by lumag** [comment]
@calebccff here is (an excerpt) of the script that I have here for the relays and PPPS hub:
```sh
#!/bin/sh

#set -x
set -e

if [ "$#" != 3 ]
then
	echo "Usage: $0 [board] [command] [on/off]"
	exit 1
fi

echo "Called $1 $2 $3"

case "$1" in 
	"rb5-2" )
		HUBLOC="10-1"
		HUBPORT="1"
		LAURENT="REL,1"
		;;
	"db820c-1" )
		HUBLOC="10-1"
		HUBPORT="2"
		LAURENT="REL,2"
		;;
	"rb1-1" )
		HUBLOC="10-1"
		HUBPORT="3"
		LAURENT="REL,3"
		;;

	* )
		echo "Unknown board $1"
		exit 2
esac

uhubctl() {
	sudo uhubctl -l $HUBLOC -p $HUBPORT -a $1
	if [ "$1" = "off" ] ; then
		sudo udevadm trigger --action=remove /sys/bus/usb/devices/${HUBLOC}.${HUBPORT}/ || true
	fi
}

laurent() {
	curl "http://laurent.lan/cmd.cgi?psw=$LAURENTPW&cmd=$LAURENT,$1"
}

#echo "$2-$3" >> /tmp/cdba-log

case "$2-$3" in
	"power-off" )
		uhubctl off
		#sleep 1
		laurent 0
		#sleep 1
		;;
	"power-on" )
		uhubctl off
		#sleep 1
		laurent 0
		#sleep 1
		laurent 1
		;;
	"usb-on" )
		uhubctl on
		;;
	"usb-off" )
		uhubctl off
		sleep 1
		;;
	"key-power-on" | "key-power-off" )
		# ignore
		;;
	*)
		echo "Unknown command $2-$3"
		exit 2
		;;
esac
```

File: device.h (hunk 2 above)
Comment (superna9999) [reply device.h]: It’s much nicer !

**Review by superna9999 — APPROVED** [review]

**Comment by calebccff** [comment]
I hooked up my pico to control the enable pin of the buck converter I have as a battery replacement, this makes all the button wrangling unecessary as I can just erase the boot partition and the devices always boot straight to fastboot mode.

There is maybe some value in this PR for someone else, but given the amount of finesse necessary to get this to work reliably, I'm not gonna stick with it... See #33

File: device.c [hunk 3]
```diff
@@ -142,13 +161,12 @@ struct device *device_open(const char *board,
...
 {
-	device->power(device, on);
+	device_control(device, power, on);
 }
 
 static void device_key(struct device *device, int key, bool asserted)
 {
-	if (device->key)
-		device->key(device, key, asserted);
+	device_control_opt(device, key, key, asserted);
```
Comment (calebccff) [inline device.c:169]: To me this is a lot less readable, without the macro it's clear what exactly this code does, but now I would have to go and find the definition of `device_control_opt` and understand it.

What about:

```c
if (device_has_control(key))
    device->key(device, key, asserted);
```

Comment (lumag) [reply device.c:169]: It becomes:
```c
if (device_has_control(key))
    device->control_ops->key(device, key, asserted);
```

The latter part was what @superna9999 disliked, too many indirections.

Comment (lumag) [reply device.c]: @calebccff reworking this to be more obvious.

**Review by superna9999 — APPROVED** [review]

**Review by konradybcio — APPROVED** [review]

**Pull request merged**